src/file_system/mod.rs

```diff
@@ -0,0 +1,345 @@
+            let mut vec = middle.to_vec();
+            vec.insert(0, first.clone());
+
+            (vec, last.clone())
+        }
+    }
+
+    pub fn from_labels(root: Label, labels: &[Label]) -> Path {
+        let mut path = NonEmpty::new(root);
+        labels.iter().cloned().for_each(|l| path.push(l));
+        Path(path)
+    }
+}
+
+pub trait RepoBackend
+where
+    Self: Sized,
+{
+    fn new() -> Directory<Self>;
+}
+
+#[derive(Debug, Clone)]
+pub enum DirectoryContents<Repo> {
+    SubDirectory(Box<Directory<Repo>>),
+    File(File),
+    Repo(Repo),
+}
+
+impl<Repo> DirectoryContents<Repo> {
+    pub fn sub_directory(directory: Directory<Repo>) -> Self {
+        DirectoryContents::SubDirectory(Box::new(directory))
+    }
+
+    pub fn file(filename: Label, contents: String) -> Self {
+        DirectoryContents::File(File { filename, contents })
+    }
+
+    pub fn repo(repo: Repo) -> Self {
+        DirectoryContents::Repo(repo)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Directory<Repo> {
+    pub label: Label,
+    pub entries: NonEmpty<DirectoryContents<Repo>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct File {
+    pub filename: Label,
+    pub contents: String,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum SystemType {
+    IsFile,
+    IsDirectory,
+}
+
+impl SystemType {
+    pub fn is_file(label: Label) -> (Label, Self) {
+        (label, SystemType::IsFile)
+    }
+
+    pub fn is_directory(label: Label) -> (Label, Self) {
+        (label, SystemType::IsDirectory)
+    }
+}
+
+impl<Repo> Directory<Repo> {
+    pub fn empty_root() -> Self
+    where
+        Repo: RepoBackend,
+    {
+        Repo::new()
+    }
+
+    pub fn list_directory(&self) -> Vec<(Label, SystemType)>
+    where
+        Repo: Clone,
+    {
+        self.entries
+            .iter()
+            .cloned()
+            .filter_map(|entry| match entry {
+                DirectoryContents::SubDirectory(dir) => Some(SystemType::is_directory(dir.label)),
+                DirectoryContents::File(file) => Some(SystemType::is_file(file.filename)),
+                DirectoryContents::Repo(_) => None,
+            })
+            .collect()
+    }
+
+    pub fn find_file(&self, path: Path) -> Option<File>
+    where
+        Repo: Clone + std::fmt::Debug,
+    {
+        let (path, filename) = path.split_last();
+        let path = NonEmpty::from_slice(&path);
+
+        let search_directory = match path {
+            None => Some(self.clone()),
+            Some(p) => self.find_directory(Path(p)),
+        };
+
+        search_directory.and_then(|dir| dir.file_in_directory(&filename))
+    }
+
+    pub fn find_directory(&self, path: Path) -> Option<Self>
+    where
+        Repo: Clone,
+    {
+        let mut search_directory = Some(self.clone());
+        let (label, labels) = path.split_first();
+        if *label == self.label {
+            for label in labels {
+                match search_directory {
```

Comment:
Something about this is funky - we know the value of `search_directory` before we start the loop, but we're checking it inside the loop. :thinking:

Reply:
I guess it's the `None` branch that seems out of place, because we're not using the `label` at all to determine whether to take that branch.

Reply:
Ya, so it's a search over the `labels`. We keep digging down into sub-directories, updating `search_directory` each time. If we don't find it then we bail out with `return None`, otherwise we keep going. If we reach the end of the labels, we'll either have found our directory or not. 

1) Would it help to add comments?
2) Can you see a cleaner way of writing it? :grimacing:

Reply:
Maybe `try_fold` over the tail of the `path`, until `get_sub_directory` returns `None`?